tests/pkg_openwsn/udp.c

```diff
@@ -0,0 +1,195 @@
+
+        /* don't run if not in synch */
+        if (ieee154e_isSynch() == FALSE) {
+            puts("Error: Node not in sync, exit");
+            return 1;
+        }
+
+        /* don't run on dagroot */
+        if (idmanager_getIsDAGroot()) {
+            puts("Error: Node is DAGROOT, exit");
+            return 1;
+        }
+
+        bool foundNeighbor = icmpv6rpl_getPreferredParentEui64(&parentNeighbor);
+        if (foundNeighbor==FALSE) {
+            puts("Error: No preferred parent EUI64, exit");
+            return 1;
+        }
+
+        /* get a free packet buffer */
+        pkt = openqueue_getFreePacketBuffer(COMPONENT_UINJECT);
+        if (pkt == NULL) {
+            puts("Error: could not create packet buffer, exit");
+            return 1;
+        }
+
+        pkt->owner = COMPONENT_UINJECT;
+        pkt->creator = COMPONENT_UINJECT;
+        pkt->l4_protocol = IANA_UDP;
+        pkt->l4_destination_port = atoi(port_str);
+        pkt->l4_sourcePortORicmpv6Type = uinject_vars.port;
+        pkt->l3_destinationAdd.type = ADDR_128B;
+        memcpy(&pkt->l3_destinationAdd.addr_128b[0], (void *)&addr, 16);
+        /* add payload */
+        packetfunctions_reserveHeaderSize(pkt, data_len);
+        memcpy(&pkt->payload[0], data, data_len);
+
+        packetfunctions_reserveHeaderSize(pkt, sizeof(uint16_t));
+        pkt->payload[1] = (uint8_t)((counter & 0xff00) >> 8);
+        pkt->payload[0] = (uint8_t)(counter & 0x00ff);
+        counter++;
+
+        packetfunctions_reserveHeaderSize(pkt, sizeof(asn_t));
+        ieee154e_getAsn(asnArray);
+        pkt->payload[0] = asnArray[0];
+        pkt->payload[1] = asnArray[1];
+        pkt->payload[2] = asnArray[2];
+        pkt->payload[3] = asnArray[3];
+        pkt->payload[4] = asnArray[4];
+
+        scheduler_push_task(push_pkt_cb, TASKPRIO_COAP);
+
+        ztimer_sleep(ZTIMER_USEC, delay);
+    }
+    return 0;
+}
+
+int udp_cmd(int argc, char **argv)
+{
+    if (argc < 2) {
+        printf("usage: %s [send|server]\n", argv[0]);
+        return 1;
+    }
+
+    if (strcmp(argv[1], "send") == 0) {
+        uint32_t num = 1;
+        uint32_t delay = 1000000LU;
+        /* don't send as root */
+        if (idmanager_vars.isDAGroot) {
+            puts("Error: Node is root, exit");
+            return 1;
+        }
+        if (argc < 5) {
+            printf("usage: %s send <addr> <port> <hex data> [<num> [<delay in us>]]\n",
+                   argv[0]);
+            return 1;
+        }
+        if (argc > 5) {
+            num = atoi(argv[5]);
+        }
+        if (argc > 6) {
+            delay = atoi(argv[6]);
+        }
+        return udp_send(argv[2], argv[3], argv[4], num, delay);
+    }
+    else if (strcmp(argv[1], "server") == 0) {
+        if (argc < 3) {
+            printf("usage: %s server [start|stop|list]\n", argv[0]);
+            return 1;
+        }
+        if (strcmp(argv[2], "start") == 0) {
```

Comment:
The "stop" command of `udp server` is not handled here. I think it should be added or removed from the usage string (the easiest ;) ).